osu.Server.Spectator/Hubs/MultiplayerHub.cs

```diff
@@ -0,0 +1,358 @@
+                        await Clients.Group(GetGroupId(room.RoomID)).MatchStarted();
+
+                        await changeRoomState(room, MultiplayerRoomState.Playing);
+                    }
+
+                    break;
+
+                case MultiplayerRoomState.Playing:
+                    if (room.Users.All(u => u.State != MultiplayerUserState.Playing))
+                    {
+                        foreach (var u in room.Users)
+                            u.State = MultiplayerUserState.Results;
+
+                        await changeRoomState(room, MultiplayerRoomState.Open);
+                        await Clients.Group(GetGroupId(room.RoomID)).ResultsReady();
+                    }
+
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Changes the provided room's state and notifies all users.
+        /// </summary>
+        private async Task changeRoomState(MultiplayerRoom room, MultiplayerRoomState newState)
+        {
+            room.State = newState;
+            await Clients.Group(GetGroupId(room.RoomID)).RoomStateChanged(newState);
+        }
+
+        /// <summary>
+        /// Given a room and a state transition, throw if there's an issue with the sequence of events.
+        /// </summary>
+        /// <param name="room">The room.</param>
+        /// <param name="oldState">The old state.</param>
+        /// <param name="newState">The new state.</param>
+        private void ensureValidStateSwitch(MultiplayerRoom room, MultiplayerUserState oldState, MultiplayerUserState newState)
+        {
+            switch (newState)
+            {
+                case MultiplayerUserState.Idle:
+                    // any state can return to idle.
+                    break;
+
+                case MultiplayerUserState.Ready:
+                    if (oldState != MultiplayerUserState.Idle)
+                        throw new InvalidStateChange(oldState, newState);
+
+                    break;
+
+                case MultiplayerUserState.WaitingForLoad:
+                    // playing state is managed by the server.
+                    throw new InvalidStateChange(oldState, newState);
+
+                case MultiplayerUserState.Loaded:
+                    if (oldState != MultiplayerUserState.WaitingForLoad)
+                        throw new InvalidStateChange(oldState, newState);
+
+                    break;
+
+                case MultiplayerUserState.Playing:
+                    // playing state is managed by the server.
+                    throw new InvalidStateChange(oldState, newState);
+
+                case MultiplayerUserState.FinishedPlay:
+                    if (oldState != MultiplayerUserState.Playing)
+                        throw new InvalidStateChange(oldState, newState);
+
+                    break;
+
+                case MultiplayerUserState.Results:
+                    if (oldState != MultiplayerUserState.Playing)
```

Comment:
Shouldn't this be checking `FinishedPlay`, or just blocking this on the client side entirely? I guess it'll be more apparent after a client implementation lands...

Reply:
good catch, should be denied. i changed how the flow worked halfway through.